### PR TITLE
added tests for custom cron sync plan 6.5 feature

### PR DESCRIPTION
**Pull request by omkarkhatavkar**
These are some tests added for custom cron options added in Sync Plan.

**Test Result :**  

```
(env) [root@okhatavk robottelo]# pytest tests/foreman/api/test_syncplan.py -v -k "custom_cron"
============================================================================ test session starts ============================================================================
platform linux -- Python 3.4.8, pytest-3.6.1, py-1.7.0, pluggy-0.6.0 -- /home/okhatavk/Satellite/robottelo/env/bin/python3.4
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/okhatavk/Satellite/robottelo, inifile:
plugins: xdist-1.23.2, services-1.3.0, mock-1.10.0, forked-0.2, cov-2.6.0
collecting 35 items                                                                                                                                                         2018-10-18 19:17:12 - conftest - DEBUG - BZ deselect is disabled in settings

collected 35 items / 32 deselected                                                                                                                                          

tests/foreman/api/test_syncplan.py::SyncPlanUpdateTestCase::test_positive_update_interval_custom_cron <- robottelo/decorators/__init__.py PASSED                      [ 33%]
tests/foreman/api/test_syncplan.py::SyncPlanProductTestCase::test_positive_add_remove_products_custom_cron <- robottelo/decorators/__init__.py PASSED                 [ 66%]
tests/foreman/api/test_syncplan.py::SyncPlanDeleteTestCase::test_positive_delete_synced_product_custom_cron <- robottelo/decorators/__init__.py PASSED                [100%]

================================================================== 3 passed, 32 deselected in 8.96 seconds ==================================================================
(env) [root@okhatavk robottelo]# pytest tests/foreman/cli/test_syncplan.py -v -k "custom_cron"
============================================================================ test session starts ============================================================================
platform linux -- Python 3.4.8, pytest-3.6.1, py-1.7.0, pluggy-0.6.0 -- /home/okhatavk/Satellite/robottelo/env/bin/python3.4
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: /home/okhatavk/Satellite/robottelo, inifile:
plugins: xdist-1.23.2, services-1.3.0, mock-1.10.0, forked-0.2, cov-2.6.0
collecting 19 items                                                                                                                                                         2018-10-18 19:17:28 - conftest - DEBUG - BZ deselect is disabled in settings

collected 19 items / 18 deselected                                                                                                                                          

tests/foreman/cli/test_syncplan.py::SyncPlanTestCase::test_positive_synchronize_custom_product_custom_cron <- robottelo/decorators/__init__.py PASSED                 [100%]

================================================================== 1 passed, 18 deselected in 6.79 seconds ==================================================================
(env) [root@okhatavk robottelo]# 

```

### CODE REVIEW

File: tests/foreman/api/test_syncplan.py [hunk 1]
```diff
@@ -1127,3 +1202,30 @@ def test_positive_delete_synced_product(self):
         sync_plan.delete()
         with self.assertRaises(HTTPError):
             sync_plan.read()
+
+    @stubbed
```
Comment (ntkathole) [inline tests/foreman/api/test_syncplan.py]: stubbed??

File: tests/foreman/api/test_syncplan.py [hunk 2]
```diff
@@ -574,6 +618,37 @@ def test_positive_repeatedly_add_remove(self):
             syncplan.remove_products(data={'product_ids': [product.id]})
             self.assertEqual(len(syncplan.read().product), 0)
 
+    @stubbed
```
Comment (ntkathole) [inline tests/foreman/api/test_syncplan.py]: stubbed??

File: tests/foreman/api/test_syncplan.py [hunk 3]
```diff
@@ -381,6 +398,33 @@ def test_positive_update_interval(self):
                     interval
                 )
 
+    @stubbed
```
Comment (ntkathole) [inline tests/foreman/api/test_syncplan.py]: same

File: tests/foreman/api/test_syncplan.py [hunk 4]
```diff
@@ -381,6 +397,32 @@ def test_positive_update_interval(self):
...
+                description=gen_string('alpha'),
+                organization=self.org,
+                interval=interval
+            ).create()
+
+            sync_plan.interval = 'custom cron'
+            sync_plan.cron_expression = gen_choice(valid_cron_expressions())
+            self.assertEqual(
+                sync_plan.update(['interval', 'cron_expression']).interval,
+                'custom cron'
```
Comment (ldjebran) [inline tests/foreman/api/test_syncplan.py]: perhaps we should add "custom cron" to SYNC_INTERVAL constant and use that.
constants.py
```python
...
SYNC_INTERVAL = {
    'hour': "hourly",
    'day': "daily",
    'week': "weekly",
    'custom': "custom cron",
}
...
```
and use it as SYNC_INTERVAL['custom']

File: tests/foreman/api/test_syncplan.py [hunk 5]
```diff
@@ -1127,3 +1199,29 @@ def test_positive_delete_synced_product(self):
...
+        sync_plan = entities.SyncPlan(
+            organization=self.org,
+            interval='custom cron',
+            cron_expression=gen_choice((valid_cron_expressions()))).create()
+        product = entities.Product(organization=self.org).create()
+        entities.Repository(product=product).create()
+        sync_plan.add_products(data={'product_ids': [product.id]})
+        product.sync()
+        sync_plan.delete()
+        self.assertEqual(None, product.sync_plan.id)
```
Comment (ldjebran) [inline tests/foreman/api/test_syncplan.py]: is this really working without reading the product again, as per my understanding product.sync_plan should be None.
eg we have to read the product twice one time after associating the sync plan and assert that product sync plan exist, and one time after deleting the sync plan and assert that product sync plan does not exist, and after try to read the sync plan it self that should raise the expected HTTP error. 
what do you think?

File: tests/foreman/api/test_syncplan.py (hunk 4 above)
Comment (omkarkhatavkar) [reply tests/foreman/api/test_syncplan.py]: make sense! will update accordingly

File: tests/foreman/api/test_syncplan.py (hunk 5 above)
Comment (omkarkhatavkar) [reply tests/foreman/api/test_syncplan.py]: agreed, done!

File: tests/foreman/api/test_syncplan.py [hunk 6]
```diff
@@ -1121,9 +1193,38 @@ def test_positive_delete_synced_product(self):
         """
         sync_plan = entities.SyncPlan(organization=self.org).create()
         product = entities.Product(organization=self.org).create()
+        entities.Repository().create()
```
Comment (ldjebran) [inline tests/foreman/api/test_syncplan.py]: perhaps:
```python
 entities.Repository(product=product).create()
```

File: tests/foreman/api/test_syncplan.py [hunk 7]
```diff
@@ -1121,9 +1193,38 @@ def test_positive_delete_synced_product(self):
...
+        product.sync()
+        sync_plan.delete()
+        with self.assertRaises(HTTPError):
+            sync_plan.read()
+
+    @run_only_on('sat')
+    @upgrade
+    def test_positive_delete_synced_product_custom_cron(self):
+        """Create a sync plan with custom cron with one synced
+        products and delete it.
```
Comment (ldjebran) [inline tests/foreman/api/test_syncplan.py]: product

File: tests/foreman/api/test_syncplan.py (hunk 5 above)
Comment (ldjebran) [reply tests/foreman/api/test_syncplan.py]: any way think there is a bug in nailgun product read  if sync_plan is null in server response we should not have sync_plan with id product sync plan must be None.

Comment (omkarkhatavkar) [reply tests/foreman/api/test_syncplan.py]: updated this to product.sync_plan for now

File: tests/foreman/api/test_syncplan.py [hunk 8]
```diff
@@ -1127,3 +1199,32 @@ def test_positive_delete_synced_product(self):
         sync_plan.delete()
         with self.assertRaises(HTTPError):
             sync_plan.read()
+
+    @run_only_on('sat')
+    @upgrade
+    def test_positive_delete_synced_product_custom_cron(self):
```
Comment (ntkathole) [inline tests/foreman/api/test_syncplan.py]: tier missing?

File: tests/foreman/cli/test_syncplan.py [hunk 9]
```diff
@@ -119,6 +120,22 @@ def valid_name_interval_update_tests():
     ]
 
 
+@filtered_datapoint
+def valid_cron_expressions():
```
Comment (ntkathole) [inline tests/foreman/cli/test_syncplan.py]: Not blocker but this looks like duplicate code.. rather we can have this list in `constant.py`

File: tests/foreman/api/test_syncplan.py (hunk 8 above)
Comment (omkarkhatavkar) [reply tests/foreman/api/test_syncplan.py]: thanks, nice catch! will update this

File: tests/foreman/cli/test_syncplan.py (hunk 9 above)
Comment (ldjebran) [reply tests/foreman/cli/test_syncplan.py]: we can put it in datafactory.py module instead, but have to write a unittest

Comment (san7ket) [reply tests/foreman/cli/test_syncplan.py]: Agree with @ldjebran , I see another use case for scap where we could use this, if we have it in datafactory

Comment (omkarkhatavkar) [reply tests/foreman/cli/test_syncplan.py]: Done!

**Pull request merged**